### ISSUE TITLE
Speed up hex conversion

### ISSUE THREAD

**Issue by gunnarvoet** [gunnarvoet/ctdproc#1]
The conversion from hex to ascii data is pretty slow. Is there any way to speed it up?

**Comment by TilmanJimenez** [comment]
Here are two changes that should improve speed (and are equivalent under the assumption hex_str is no longer than 6)
```
def _hexword2freq(self, hex_str):
    return int(hex_str, 16) / 256
```
which I measured to be 4 times quicker, and
```
def _hexword2volt(self, hex_str):
    e = int(hex_str, 16) ^ 0xffffff
    v1 = (e >> 12)/891
    v2 = (e & 0xfff)/891
    return v1, v2
```
which should be 5.5 times faster. The second function is the same to 10 decimal points (but due to how floats are handled, it might round differently at the 11th), which for all measurement purposes this should not be relevant.
Since these functions are called up to 5 and 4 times per measurement, it should help speed up at least a bit.

**Comment by gunnarvoet** [comment]
Hi @TilmanJimenez, thanks so much for your comment! I tried your suggestions - the new `_hexword2freq` function appears to work well and does speed up the processing. `_hexword2volt` results in non-negligible changes to the output. Unfortunately, I am not very versed in hex data so can't dive into what might be wrong there.

The test directory has a couple example data files if you wanted to play with it.

**Comment by TilmanJimenez** [comment]
Hey @gunnarvoet sorry I messed up!
The division is supposed to be by 819 not 891. I had written a test over all 16⁶ possibilities to make sure there was no difference, so I was really confused as to what happened, but apparently I twisted the numbers when submitting the comment.

A quick explanation into what happens: The first one is quite easy: 0x12 "shifted" 2 to the left (so 0x1200) is just multiplying by 16² = 256. So instead of converting every hex pair into an int and then multiplying/dividing, I just divide `0x123456` by 256 which would be `(0x12*256²+0x34*256+0x56)/256`

The second one is a bit more complicated but not that much. First `5*(1-v1/4095) = 5*(4095-v1)/4095 = (4095-v1)/819`
Then 4095 in binary would be `0b111111111111`, which is really practical! If we see v1 as a 12 bit long binary number as well (as an example `0b000010100100`). Then when we subtract, everywhere where there is a 1 in v1, 4095-v1 will have a 0 and everywhere where v1 has a 0, 4095-v1 will have a 1:

 ```
  111111111111
-000010100100
  ------------
  111101011011
```

While technically we could just NOT v1, this won't work in Python. So the easiest solution is to XOR with `0xfff` (4095).
But we don't need to perform the "subtraction" on each of the two numbers separately!
```
 0xfff|fff
^0x123|456
---------
 0xedc|ba9
```
So after converting the 6 hex numbers into an int, we can XOR them with `0xffffff`. Then we just need to extract the numbers from the result. The first 3 hex numbers (for v1) will we the result if we shift the last 3 away (so in this case ba9), with every hex number being 4 bits, moving by 3 hex numbers is a 12 bit shift (`e>>12`). And the second 3 are just the last 12 bits of the number, so ANDing with `0xfff` will cut the rest away.
Then we calculated 4095-v1 and 4095-v2, so we just need to divide by 819, and we're done :)

I'm not sure how much time I will have for this project in the following weeks, but if and when I do I'll clone the repository so I can submit changes directly!
